lib/puppet/pops/types/type_calculator.rb

```diff
@@ -452,9 +452,18 @@ def instance_of_PTupleType(t, o)
 
   def instance_of_PStructType(t, o)
     return false unless o.is_a?(Hash)
-    h = t.hashed_elements
-    # all keys must be present and have a value (even if nil/undef)
-    (o.keys - h.keys).empty? && h.all? { |k,v| instance_of(v, o[k]) }
```

Comment:
I don't get why this did not work. The first part (o.keys - h.keys) should be empty if o has fewer keys than h. The second part where instance_of(v, o[k]) should end up checking if nil is instance of `Optional[T]` which it should be.